lib/saluki-io/src/compression/zstd.rs

```diff
@@ -0,0 +1,757 @@
+                    panic!("should not flush to the output buffer during the write phase")
+                }
+                Err(e) => panic!("write_all failed: {}", e),
+            }
+        }
+
+        // Now trigger our flush and make sure we fill up the output buffer.
+        //
+        // Make sure our output buffer is still empty before doing so.
+        assert_eq!(output.as_slice().len(), 0);
+
+        match cctx.flush(&mut output) {
+            Ok(WriteStatus::Ok) => panic!("should not be able to fully flush to the output buffer"),
+            Ok(WriteStatus::OutputBufferFull) => {
+                // We have a full output buffer, so let's check the status.
+                assert_eq!(output.as_slice().len(), output.capacity());
+            }
+            Err(e) => panic!("flush failed: {}", e),
+        }
+    }
+
+    #[test]
+    fn simple_cctx_output_buffer_full_finalize() {
+        // Create a simple compression context and an output buffer with a very low capacity.
+        //
+        // Our goal is to write a bunch of data into the compression context and then finalize the compressor, and then
+        // assert that the call returns `WriteStatus::OutputBufferFull` once it fills it all up.
+        let mut cctx = create_simple_cctx().unwrap();
+
+        let mut output_buffer = vec![0; 8];
+        let mut output = OutBuffer::around_pos(&mut output_buffer, 0);
+
+        for i in 0..4 {
+            // A simple input string with some different data in it. Small enough that it should all be buffered
+            // internally and not trigger an implicit flush when writing... but will lead to enough output when we call
+            // flush that it will fill up the output buffer.
+            let input_data = format!("{} my {} special {} string", i, i, i);
+            let mut input = InBuffer::around(b"hello world!");
+
+            match cctx.write_all(&mut input, &mut output) {
+                Ok(WriteStatus::Ok) => continue,
+                Ok(WriteStatus::OutputBufferFull) => {
+                    panic!("should not flush to the output buffer during the write phase")
+                }
+                Err(e) => panic!("write_all failed: {}", e),
+            }
+        }
+
+        // Now trigger our finalize and make sure we fill up the output buffer.
+        //
+        // Make sure our output buffer is still empty before doing so.
+        assert_eq!(output.as_slice().len(), 0);
+
+        match cctx.finalize(&mut output) {
+            Ok(WriteStatus::Ok) => panic!("should not be able to fully finalize to the output buffer"),
+            Ok(WriteStatus::OutputBufferFull) => {
+                // We have a full output buffer, so let's check the status.
+                assert_eq!(output.as_slice().len(), output.capacity());
+            }
+            Err(e) => panic!("finalize failed: {}", e),
+        }
+    }
+
+    #[test_strategy::proptest(async = "tokio")]
```

Comment:
We use this non-traditional way of creating a `proptest`-based test as `proptest`'s normal `proptest!` declarative macro _cannot_ be used when wrapped around an async test function that uses `#[tokio::test]`.